### PR TITLE
fix(engine+cli): GPT-2 byte map, incremental UTF-8 streaming, daemon reset, serve rewrite

### LINKED ISSUE #8: [gfx1100][ROCm 6.4] v0.1.2-alpha: shared output corruption, severe context-length degradation, confirmed cross-request state bleed, precompiled kernel issues
Hi — I did a full diagnostic pass on **hipfire v0.1.2-alpha** on **RX 7900 XTX (gfx1100, 24GB)** and compared it against **v0.1.0** on the same machine.

This report focuses on **core runtime / output correctness / serve isolation**.
I'm filing a separate PR for CLI/API fixes.

---

## Environment

| Component | Value |
|-----------|-------|
| GPU | AMD Radeon RX 7900 XTX |
| Arch | gfx1100 |
| VRAM | 24 GB |
| CPU | Ryzen 9 9950X3D |
| OS | openSUSE Tumbleweed |
| ROCm | 6.4 |
| hipfire | v0.1.2-alpha |
| Comparison | v0.1.0 |

---

### Confirmed improvements
- `serve` is much more stable than in v0.1.0
- 27B works via serve
- Sequential request handling is stable
- Stream cancel + recovery works

### Confirmed problems
1. **27B HFQ4 throughput regression** vs v0.1.0
2. **Shared output corruption symptoms across HFQ4 and HFQ6**
3. **Severe context-length degradation** on 27B HFQ6
4. **Cross-request state bleed in concurrent serve**
5. **Precompiled gfx1100 kernels produce garbage output** on my setup
6. **Prompt escape handling is incorrect** (`\n`/`\t` not interpreted)
7. **`<think>` tokens count toward `--max-tokens`**

---

## Benchmark summary

| Model / Quant | v0.1.0 | v0.1.2-alpha | Delta |
|---------------|--------|-------------|-------|
| 9B HFQ4 | ~62 tok/s | 58–61 tok/s | roughly flat |
| 27B HFQ4 | 25–27 tok/s | 20–22 tok/s | **~20% regression** |
| 27B HFQ6 | 16–18 tok/s | 16–18 tok/s | stable |
| llama.cpp Qwen3.5-27B Q5_K_M | 15–18 tok/s | 15–18 tok/s | reference |

---

### 1. Determinism test

**Setup:** temp=0, same prompt, 5 runs

**Result:** Output is **identical across runs**. Problems are systematic, not random.

---

### 2. Exact-copy test

**Prompt:** `Repeat exactly: alpha beta\ngamma\tdelta`

| Model | Output | Verdict |
|-------|--------|---------|
| 9B | `alpha betagammatdelta` | ❌ collapsed |
| 27B HFQ4 | `alpha beta gamma delta` | ⚠️ normalized to spaces |
| 27B HFQ6 | `alpha beta gamma delta` | ⚠️ normalized to spaces |

**Expected:** Preserved separators. **Observed:** Lost or normalized.

---

### 3. Prime numbers test

**Prompt:** First 5 prime numbers, comma-separated, temp=0

| Model | Output (identical every run) |
|-------|------------------------------|
| 27B HFQ4 | `2,3,5711` |
| 27B HFQ6 | `2,3,5711` |

**Expected:** `2, 3, 5, 7, 11`

Identical failure on both quants → **shared code path issue**, not quant-specific.

---

### 4. Context-length sweep (27B HFQ6)

| Prompt length | tok/s | Quality |
|---------------|-------|---------|
| Short (~20 tok) | 14.9 | Passable |
| Medium (~200 tok) | 11.2 (-25%) | Degrades to syntax soup |
| Long (~500 tok) | 5.2 (**-65%**) | Garbage / rambling |

**This is the most serious runtime issue.** Makes 27B unusable for realistic tasks.

Consistent with **context-length-sensitive state corruption** (KV/cache/recurrent state).

---

### 5. Cross-request state bleed (CONFIRMED)

**Repro:** Two concurrent requests:

| Prompt | Expected | Observed |
|--------|----------|----------|
| "Capital of France" | Paris | `The user wants to know the answer.` |
| "Capital of Japan" | Tokyo | `<think>\n\n</think><\|endoftext\|>` |

**Confirmed cross-request state bleed / missing session isolation.**

---

### 6. Structured output tests

| Test | HFQ4 | HFQ6 |
|------|-------|------|
| JSON `{"a":1,"b":2}` | ❌ mixed escaped/unescaped | ⚠️ OK then rambles |
| Case preservation | ❌ `snake_camelC_PascalCASE_UPPER_lower_` | ⚠️ minor errors |
| Numbers 1-10 | ❌ `234567890` | ❌ `123456789` |
| fizzbuzz | ❌ garbled syntax | ❌ regression from v0.1.0 |
| binary search | ❌ missing comparison value | ❌ same |

---

### 7. Single-token test

| Prompt | Output | Verdict |
|--------|--------|---------|
| Return exactly: 42 | 42 | ✅ |
| Return exactly: yes | yes | ✅ |
| Return exactly: , | , | ✅ |
| Return exactly: 100 | 1024 | ❌ hallucination |

Short outputs are mostly correct → corruption is **not from the first token** but worsens with generation length.

---

### 8. Prompt escape handling

| Input | Output | Issue |
|-------|--------|-------|
| Literal newlines | `1 2 3` | newlines → spaces |
| Escaped `\n` | `1n2n3` | backslash eaten, n kept |
| Tabs | `A B C` | tabs → spaces |

CLI does not interpret escape sequences as expected.

---

### 9. `<think>` tags consume `--max-tokens`

`<think>\n\n</think>` uses ~5 tokens from the user-visible budget. With `--max-tokens 5`, only 1-2 content tokens remain.

A `--no-think` flag or excluding think tokens from the budget would help.

---

## Suspected root causes

| # | Hypothesis | Confidence | Key evidence |
|---|-----------|------------|-------------|
| A | Shared output assembly / decode-path issue | High | `2,3,5711` identical on both quants |
| B | Context-length-sensitive state corruption | High | -65% speed + quality collapse at ~500 tok |
| C | Cross-request state bleed in serve | High | Confirmed concurrent interference |
| D | HFQ4-specific performance regression | Medium-High | 20% drop, HFQ6 unaffected |
| E | Precompiled gfx1100 kernel correctness issue | High | Garbage with precompiled, fine with JIT |

---

## Symptom → likely layer

| Symptom | Suspected layer |
|---------|----------------|
| Collapsed whitespace / missing separators | Decode / postprocess / token assembly |
| Identical bug on both quants | Shared path, not quant-specific |
| Long-context collapse | KV / state / cache / attention path |
| HFQ4-only perf regression | HFQ4 dequant / matmul / scheduling |
| Concurrent garbling | Serve / session isolation |

---

### P0
- Cross-request state bleed in concurrent serve
- Precompiled gfx1100 kernels producing garbage output

### P1
- Context-length-sensitive degradation / state corruption
- Shared output assembly / decode-path corruption

### P2
- HFQ4 throughput regression

### P3
- Prompt escape handling
- `<think>` counting toward max_tokens
- Remaining API polish

---

## Minimal repro prompts

```
# Prompt 1 — exact copy
Repeat exactly: alpha beta\ngamma\tdelta

# Prompt 2 — primes
Return the first 5 prime numbers, comma-separated.

# Prompt 3 — numbers
Return numbers 1 to 10, one per line.

# Prompt 4 — JSON
Return exactly this JSON: {"a":1,"b":2}

# Prompt 5 — concurrent pair (run simultaneously)
Request A: What is the capital of France?
Request B: What is the capital of Japan?
```

---

## What I can provide

- Exact prompts and raw outputs
- `hipfire diag` output
- Before/after CLI/API results
- Reproductions for concurrent serve behavior
- CLI patch (separate PR)

### PR DESCRIPTION
**Pull request by domvox**
## fix(engine+cli): GPT-2 byte map, incremental UTF-8 streaming, daemon reset, serve rewrite

Fixes #8 (remaining items from v0.1.2 diagnostic report)

### What this PR does

**7 fixes across 3 files**, tested on RX 7900 XTX (gfx1100, 24GB, ROCm 6.4).

### 1. Fix GPT-2 byte-to-unicode mapping (`tokenizer.rs`)

`gpt2_char_to_byte()` used `byte + 256` for non-printable bytes, but GPT-2's `bytes_to_unicode()` assigns sequential offsets. Byte 0x8C mapped to U+018C instead of U+012E, causing `from_utf8_lossy` to produce `U+FFFD` artifacts.

**Before:** `2�.3�.5�.7�.11` / `�+lysée`
**After:** `2，3，5،7،11` / `Île de la Cité`

Also changed `decode()` to collect all bytes before UTF-8 conversion — individual tokens in Qwen's byte-level BPE can be incomplete UTF-8 sequences.

### 2. Incremental UTF-8 streaming (`daemon.rs`)

Replaced per-token `tokenizer.decode(&[next_token])` with buffered `decode_bytes()` + `from_utf8` with `valid_up_to()`. Fixes panic on multi-byte characters (emoji like 😊 span 4 bytes across token boundaries).

### 3. Daemon `reset` IPC command (`daemon.rs`)

Added `{"type":"reset"}` command that zeros DeltaNet S matrices, scales, conv states, and KV cache K/V buffers via `gpu.hip.memset`. Returns `{"type":"reset","ok":true}`.

### 4. Default Q8 KV (`cli/index.ts`)

Changed hardcoded `turbo: 4` to `turbo: process.env.TURBO ? Number(process.env.TURBO) : 0`. TurboQuant is broken on long context for 27B — this was causing 0.3 tok/s / empty output at ~500 tok prompts. With Q8 KV, same prompt runs at 10.4 tok/s.

### 5. Serve handler rewrite (`cli/index.ts`)

- Sends `reset` to daemon before each HTTP request (fixes state bleed)
- Builds proper ChatML prompt from messages array (fixes system prompt)
- Full OpenAI `chat.completion` / `chat.completion.chunk` response format
- Single `safeRelease()` lock wrapper (fixes double releaseLock)
- Passes `top_p`, `repeat_penalty` to daemon

### 6. Think tag stripping (`cli/index.ts`)

Filters `<think>...</think>` blocks and `<|im_end|>` tokens in both `run` and `serve` output. Streaming filters in real-time; non-streaming uses regex.

### 7. Serve pre-warm (`cli/index.ts`)

Loads default model and generates 1 warmup token before opening HTTP port. Eliminates cold-start timeout crash (JIT kernel compilation takes 10-30s on first run).

### Test results

| Test | Before | After |
|------|--------|-------|
| `U+FFFD` artifacts | `2�.3�.5...` | `2，3，5...` ✅ |
| Emoji streaming | crash (panic) | works ✅ |
| Long context 27B HFQ4 | 0.3 tok/s (turbo) | 10.4 tok/s (Q8) ✅ |
| Serve state bleed | echoes prev request | independent ✅ |
| Serve system prompt | ignored | works ✅ |
| Serve OpenAI fields | missing | full schema ✅ |
| Think tags in output | visible | stripped ✅ |
| Serve cold start | timeout crash | 3s pre-warm ✅ |
| 9B speed | 58-61 tok/s | 65.1 tok/s ✅ |
| 27B speed | 20-22 tok/s | 23.5 tok/s ✅ |

### Files changed

```
crates/engine/src/tokenizer.rs  — GPT-2 byte map fix, decode_bytes(), batch decode
crates/engine/examples/daemon.rs — reset command, incremental UTF-8 streaming
cli/index.ts                     — turbo default, serve rewrite, think strip, pre-warm
```

### CODE REVIEW

**Review by Kaden-Schutt — CHANGES_REQUESTED** [review]
## Review: fix(engine+cli): GPT-2 byte map, incremental UTF-8 streaming, daemon reset, serve rewrite

Great work on the 7900 XTX testing — the tokenizer byte map fix and incremental UTF-8 streaming are solid and clearly needed. Several of these changes should land, but the CLI rewrite has conflicts with master that need resolving first.

---

### Approve (cherry-pick ready)

**tokenizer.rs — GPT-2 byte map fix**: Correct. The `bytes_to_unicode()` tables with `GPT2_BYTE_TO_OFFSET`/`GPT2_OFFSET_TO_BYTE` properly implement the sequential offset mapping. The old `b + 256` was wrong for bytes 127-160. `decode_bytes()` for incremental streaming is a clean addition.

**daemon.rs — Incremental UTF-8 streaming**: Correct pattern. `valid_up_to()` handles multi-byte chars spanning token boundaries cleanly.

**turbo default to Q8**: Correct decision. TurboQuant on 27B long context is known-broken (RDNA L2 pressure, see our turbo RDNA analysis). `TURBO=N` env opt-in is the right UX.

**safeRelease lock pattern**: Better than the old double-release. `cancel()` handler on the ReadableStream is also good.

**Think tag stripping + pre-warm**: Both good additions.

---

### Must-fix before merge

#### 1. File naming regression (cli/index.ts) — **blocks merge**

The PR reverts filenames from `.hf4`/`.hf6` back to `.q4.hfq`/`.hfq6.hfq`. Master already moved to the new naming in commit `cbe5c59` and the HuggingFace repos (`schuttdev/hipfire-*`) now serve `.hf4`/`.hf6` files. This will break `hipfire pull` for all users on current master.

Similarly, `findModel()` and `listLocal()` are changed to only search `.hfq` — this won't find any `.hf4`/`.hf6` models on disk.

**Fix**: Rebase on master and keep the `.hf4`/`.hf6` naming. The backward-compat fallback in `findModel()` should stay.

#### 2. Reset handler missing seq_pos + conversation_tokens (daemon.rs)

The PR's reset zeros GPU buffers but does NOT reset:
```rust
m.seq_pos = 0;
m.conversation_tokens.clear();
```

Master already does this (lines 164-165). Without it, after reset the next `forward_scratch` call writes KV at the old `seq_pos` offset into a zeroed-out cache. Attention reads positions `0..seq_pos` — all zeros from the cleared buffers — producing garbage logits. `conversation_tokens` drives repeat penalty, so stale tokens from prior conversations will penalize the new generation.

**Fix**: Add `m.seq_pos = 0; m.conversation_tokens.clear();` to the reset handler. (Master already has this — rebase will pick it up.)

#### 3. System prompt double-wrapping (cli/index.ts serve handler)

The PR builds ChatML by concatenating all messages including system:
```ts
for (const m of body.messages || []) {
  prompt += `<|im_start|>${m.role}\n${m.content}<|im_end|>\n`;
}
```

Then passes it as `prompt` with no `system` parameter. But `generate()` in daemon.rs wraps the `prompt` in `<|im_start|>user\n...<|im_end|>` tokens (lines 346-355). This produces double-wrapped malformed ChatML:

```
<|im_start|>user
<|im_start|>system
You are a helpful assistant.
<|im_end|>
<|im_start|>user
Hello
<|im_end|>
<|im_end|>
```

**Fix**: Extract the system message and pass it via the `system` field in the generate params. Pass only the conversation content as `prompt`. This is what master's serve handler does.

#### 4. Removed tool calling support (cli/index.ts)

Commit `3096e21` on master added `parseToolCalls()` and Hermes-format tool injection for OpenAI API compatibility. The PR removes both. This breaks the Hermes agent workflow (carnice model + hermes-agent profile).

**Fix**: Keep the tool calling code from master. The serve rewrite can be merged separately from the tool handling.

#### 5. Removed community finetune models

Carnice (9b, 9b-hf6) and Qwopus (4b, 9b, 9b-hf6, 27b) entries are deleted from the registry. These models are published on HuggingFace and actively used.

**Fix**: Keep the community finetune entries. They were added in commit `be2c507`.

---

### Should-fix

#### 6. Hardcoded arch detection (cli/index.ts update)

Replaced the generic formula (`Math.floor(ver / 10000)` etc.) with a 5-entry hardcoded lookup. New GPU arch IDs (RDNA4 stepping variants, APUs, etc.) won't be detected.

**Fix**: Keep the generic formula from master — it handles all `gfx_target_version` values correctly.

---

### Suggested merge path

1. Cherry-pick **tokenizer.rs** changes directly (GPT-2 byte map + decode_bytes + incremental streaming)
2. Cherry-pick **daemon.rs** incremental UTF-8 streaming (the `streamed_tokens`/`emitted_bytes` pattern in all 3 generate paths)
3. Take the **turbo default**, **think strip**, **pre-warm**, **safeRelease** changes from cli/index.ts but rebase them on master's current index.ts (keeping .hf4/.hf6 naming, tool calling, community models, generic arch detection)

Happy to help rebase if useful. The tokenizer and streaming fixes are high-value and should land quickly.

**Comment by domvox** [comment]
Note: I've changed my GitHub username from `Sludge2158` to `domvox`. Same person, same work. The clean cherry-pick of approved changes is in PR #11.

**Comment by domvox** [comment]
Closing — tokenizer fixes landed via b831920, and the serve/daemon changes have been addressed separately in master. Thanks for the detailed review.
